game-core/src/main/java/games/strategy/triplea/delegate/battle/firinggroups/FiringGroup.java

```diff
@@ -0,0 +1,80 @@
+package games.strategy.triplea.delegate.battle.firinggroups;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.attachments.UnitAttachment;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import lombok.NonNull;
+import lombok.Value;
+
+/**
+ * Breaks list of units into groups of non suicide on hit units and each type of suicide on hit
+ * units since each type of suicide on hit units need to roll separately to know which ones get
+ * hits.
+ */
+@Value(staticConstructor = "of")
+public class FiringGroup {
+  private final @NonNull Collection<Unit> firingGroup;
+  private final @NonNull Collection<Unit> validTargets;
+  private final boolean suicide;
+  // type is currently only used by Aa firing groups
+  private final @NonNull String type;
+
+  static List<FiringGroup> newFiringUnitGroups(
+      final Collection<Unit> units,
+      final Function<Collection<Unit>, Collection<Unit>> getValidTargets,
+      final String type) {
+
+    // split the units by suicideOnHit and unitType (if suicideOnHit)
+    final Map<UnitType, Collection<Unit>> suicideUnitsByType = new HashMap<>();
```

Comment:
Any thoughts to using guava multimap? It's effectively the same data structure as a `Map<X, Collection<Y>>` and simplifies adding elements and retrieving.

That would look like this:
```
    final Multimap<UnitType, Unit> suicideUnitsByType = HashMultimap.create();
    final Collection<Unit> nonSuicideFiringGroup = new ArrayList<>();
    for (final Unit unit : units) {
      final UnitType unitType = unit.getType();
      if (UnitAttachment.get(unitType).getIsSuicideOnHit()) {
        suicideUnitsByType.put(unitType, unit);  // << note the computeIfAbsent is removed
      } else {
        nonSuicideFiringGroup.add(unit);
      }
    }
```
Then when accessing later:
```
       for (final Collection<Unit> suicideFiringGroup : suicideUnitsByType.asMap().values()) {
```

Reply:
The original code used the Map.  If you want me to switch it to guava multimap, I have no problem with that.

Reply:
Is this mostly just a straight code move? If we're doing more than simply moving code, then treating it as new code is relatively appropriate (it's arguably in effect re-written/new).

Reply:
It is mostly a straight code move with changes added that are requested by either you or @asvitkine.  I've also made changes to resolve the code climate complexities.  It is probably about half rewritten at this point.